test/persist-catalog-migration/mzcompose.py

```diff
@@ -69,8 +70,228 @@ def workflow_test_migration_and_rollback(c: Composition) -> None:
+
+    # Panic during migration.
+    try:
+        with c.override(
+            Materialized(
+                catalog_store="stash",
+                environment_extra=["FAILPOINTS=post_persist_fence=panic"],
+            )
+        ):
+            # Note: We actually want to retry this 0 times, but we need to retry at least once so a
+            # UIError is raised instead of an AssertionError
+            c.up("materialized", max_tries=1)
+    except UIError:
+        pass
+
+    # Complete migration and check objects.
+    c.up("materialized")
+    check_objects(c, 0)
+
+
+def workflow_test_rollback_panic_after_fence_then_rollback(c: Composition):
+    c.down(destroy_volumes=True)
+    c.up("testdrive", persistent=True)
+
+    # Create initial objects and update flag.
+    c.up("materialized")
+    create_objects(c, 0)
+    c.testdrive(
+        input=dedent(
+            """
+    $ postgres-execute connection=postgres://mz_system@materialized:6877/materialize
+    ALTER SYSTEM SET catalog_kind TO 'persist'
+    """
+        )
+    )
+    c.kill("materialized")
+
+    # Perform migration, check objects, create more objects, and update flag.
+    c.up("materialized")
+    check_objects(c, 0)
+    create_objects(c, 1)
+    c.testdrive(
+        input=dedent(
+            """
+    $ postgres-execute connection=postgres://mz_system@materialized:6877/materialize
+    ALTER SYSTEM SET catalog_kind TO 'stash'
+    """
+        )
+    )
+    c.kill("materialized")
+
+    # Panic during rollback.
+    try:
+        with c.override(
+            Materialized(
+                catalog_store="stash",
+                environment_extra=["FAILPOINTS=post_persist_fence=panic"],
+            )
+        ):
+            # Note: We actually want to retry this 0 times, but we need to retry at least once so a
+            # UIError is raised instead of an AssertionError
+            c.up("materialized", max_tries=1)
+    except UIError:
+        pass
+
+    # Complete migration and check objects.
+    c.up("materialized")
+    check_objects(c, 0)
+    check_objects(c, 1)
+
+
+def workflow_test_migration_panic_after_write_then_migrate(c: Composition):
+    c.down(destroy_volumes=True)
+    c.up("testdrive", persistent=True)
+
+    # Create initial objects and update flag.
+    c.up("materialized")
+    create_objects(c, 0)
+    c.testdrive(
+        input=dedent(
+            """
+    $ postgres-execute connection=postgres://mz_system@materialized:6877/materialize
+    ALTER SYSTEM SET catalog_kind TO 'persist'
+    """
+        )
+    )
+    c.kill("materialized")
+
+    # Panic during migration.
+    try:
+        with c.override(
+            Materialized(
+                catalog_store="stash",
+                environment_extra=["FAILPOINTS=migrate_post_write=panic"],
+            )
+        ):
+            # Note: We actually want to retry this 0 times, but we need to retry at least once so a
+            # UIError is raised instead of an AssertionError
```

Comment:
What is the assertion that fails?

Reply:
This one: https://github.com/MaterializeInc/materialize/blob/d5cade32235c63627433718b406f7a004531c76e/misc/python/materialize/mzcompose/composition.py#L317